README.md
# Energy Simulator:

A repos of energy simulator. 

See Also energy playground which may change name eventually.




CHANGES.txt
v0.1.0, 12/12/2016 -- Initial release.
v0.1.1, 03/01/2017 -- Moved LICENSE.txt to LICENSE per google open-source cross check.
v0.1.3, 03/08/2017 -- Removed protobuf support
v1.0.0, 06/02/2017 -- Added Transmission, GridRecStorage and Tests.


gridsim/__init__.py
# Copyright 2017 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Linear Program Constraints to simulate different energy assumptions.

This package was originally written to support understanding of energy
solutions within Google Research.  It has subsequently been used to
generate scenarios for a website

http://energystrategies.org

to help show policy makers, and other interested people, the cost
optimal way to reduce CO2 emissions from the electricity sector.  This
package has been released under open-source in order to fully document
how the results of the website were determined and to help further
understanding and analysis of energy costs.  See ../LICENSE for more
details of the release.

The reason for using this model is to avoid the limitations of
Levelized Cost of Electricity (LCOE) which is a popular metric for
computing costs within the energy sector.  The idea behind LCOE is to
find an equivalent cost between different energy sources which have
different capital, fixed and operating costs, and which have different
capacity factors (Percentage of the time when the sources can provide
power e.g. solar power during the day.)

A major limitation of LCOE is that it makes the assumption that
electricity is fungible, that a kilowatt-hour generated during the day
is equivalent to a kilowatt-hour generated at night.  This assumption
falls apart in the real world.  As of 2016, there are many articles
which say that solar is cheaper than fossil fuels.  Yet due to solar's
intermittency, other energy sources must be built to support nighttime
and cloudy day use.  This fundamentally changes the cost optimization
in a way which LCOE does not account for.  Since the other energy
sources must be built anyway to support nighttime use, then the proper
analysis would be to compare solar capital costs for power during the
day against the marginal cost of running the other energy sources
during the day.  LCOE does not consider this.  This package does.

This package does an hour by hour cost optimization of capital,
variable and operating costs of various different energy sources.  It
uses historical hour-by-hour data to determine actual overall demand
and actual power available for intermittent power sources such as
solar and wind power.

This package uses a linear program (LP) to optimize capital and
variable costs subject to the following basic constraint:

  - Total supply power must meet or exceed total demand for all hours.
      (Colloquially called 'Keeping the lights on.')

Additionally optional constaint is:

  - Renewable Portfolio Standard (RPS): Most states in the US have an
      RPS in place.  It requires that a certain percentage of the
      total power grid come from certain sources (such as wind or
      solar).  This package allows selection of RPS percentage as well
      as which sources should be in the RPS.

Power Sources specify capital, operating costs, maximum power, energy
limits and co2 emitted.  Overall the simulation may also specify a
carbon tax and discount rate of money to determine the current value
of future costs while simulating different cost assumptions.
"""


gridsim/grid_sim_simple_example.py
# Copyright 2017 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


"""Example entry-point for using grid_sim_linear_program to analyize energy.

Populates and runs the linear program for a simple case of solar power
and natural gas power.
"""

import os.path as osp

import distutils.sysconfig as sysconfig

import grid_sim_linear_program as gslp

import pandas as pd


def get_data_directory():
  """Returns a path to grid_sim data in site-lib packages."""

  abs_package_path = osp.dirname(gslp.__file__)
  return [abs_package_path, 'data']


def simple_lp(profile_dataframe):
  """Builds a simple LP with natural gas and solar.

  Args:
   profile_dataframe: A pandas dataframe with source and demand names
     in the column header.  Hour of year in indexed column-0.

  Returns:
    A LinearProgramContainer with natural gas and solar sources.
  """

  lp = gslp.LinearProgramContainer(profile_dataframe)


  # Must specify a demand to put proper load on the grid.  GridDemand
  # is linked to the corresponding profile:
  # lp.profiles[GridDemand.name]

  lp.add_demands(gslp.GridDemand('DEMAND'))

  # Nondispatchable sources are intermittent and have profiles which
  # describe their availability.  Building more capacity of these
  # sources scales the profile but cannot provide power when the
  # profile is 0. (e.g. Solar power in the middle of the night, Wind
  # power during a lull)).  A nondispatchable GridSource is linked to
  # its corresponding profile, profile[GridSource.name]

  lp.add_nondispatchable_sources(
      gslp.GridSource(
          name='SOLAR',  # Matches profile column name for nondispatch.
          nameplate_unit_cost=946000,  # Aggressive solar cost $/MW
          variable_unit_cost=0,  # No fuel cost.
          co2_per_electrical_energy=0,  # Clean energy.
          is_rps_source=True))  # In Renewable Portfolio Standard

  # Dispatchable sources can provide power at any-time.  The LP will
  # optimize generation from these sources on an as-needed basis
  # hour-by-hour.

  lp.add_dispatchable_sources(
      gslp.GridSource(
          name='NG',  # Dispatchable, so no name restriction.
          nameplate_unit_cost=1239031,  # Cost for a combined cycle plant. $/MW
          variable_unit_cost=17.5,  # Cheap fuel costs assumes fracking. $/MWh
          co2_per_electrical_energy=0.33,  # Tonnes CO2 / MWh
          is_rps_source=False))  # Not in Renewable Portfolio Standard.

  return lp


def adjust_lp_policy(lp,
                     carbon_tax=0,
                     renewable_portfolio_percentage=30,
                     annual_discount_rate=0.06,
                     lifetime_in_years=30):
  """Configures the LP based upon macro-economic costs and policy.

  Args:
    lp: LinearProgramContainer
    carbon_tax: Float cost of emitting co2 in $ / Tonne-of-CO2.
      Default is no carbon tax.

    renewable_portfolio_percentage: 0. <= Float <= 100. Percentage of
      total generation which must come from sources in the Renewable
      Portfolio Standard.  LP may not converge if this is set to a
      high amount without any storage elements in the LP.  Default is
      30%.

    annual_discount_rate: interest rate used in discounted cash flow
      analysis to determine the present value of future costs. Default
      value is 6% (0.06).

    lifetime_in_years: Float Number of years over which fuel costs
      are paid off.  Default is 30 years.
  """

  hours_per_year = 24 * 365
  lp.carbon_tax = carbon_tax
  lp.rps_percent = renewable_portfolio_percentage
  lp.cost_of_money = gslp.extrapolate_cost(
      1.0,
      annual_discount_rate,
      lp.number_of_timeslices / hours_per_year,
      lifetime_in_years)


def display_lp_results(lp):
  """Prints out costs, generation and co2 results for the lp.

  Must be run after lp.solve() is called.

  Args:
    lp: LinearProgramContainer containing sources and storage.
  """

  system_cost = 0
  system_co2 = 0

  # Loop over sources and display results.
  sources = lp.sources
  for source in sources:
    capacity = source.get_nameplate_solution_value()
    generated = sum(source.get_solution_values())
    co2 = source.co2_per_electrical_energy * generated
    capital_cost = source.nameplate_unit_cost * capacity
    fuel_cost = (source.variable_unit_cost * generated +
                 co2 * lp.carbon_tax) * lp.cost_of_money
    total_source_cost = capital_cost + fuel_cost

    print """SOURCE: %s
  Capacity: %.2f Megawatts
  Generated: %.2f Megawatt-hours
  Emitted: %.2f Tonnes of CO2
  Capital Cost: $%.2f
  Fuel Cost: $%.2f
  Total Cost: $%.2f""" % (source.name,
                          capacity,
                          generated,
                          co2,
                          capital_cost,
                          fuel_cost,
                          total_source_cost)

    system_cost += total_source_cost
    system_co2 += co2

  # Loop over storage and display results.
  for storage in lp.storage:
    capacity = storage.get_nameplate_solution_value()
    unused_stored = storage.get_solution_values()
    charge_capacity = storage.charge_nameplate.solution_value()
    discharge_capacity = storage.discharge_nameplate.solution_value()
    storage_cost = sum(
        [capacity * storage.storage_nameplate_cost,
         charge_capacity * storage.charge_nameplate_cost,
         discharge_capacity * storage.discharge_nameplate_cost])
    print """STORAGE: %s
  Capacity: %.2f Megawatt-hours
  Maximum Charge Power: %.2f Megawatts
  Maximum Discharge Power: %.2f Megawatts
  Total Cost: $%.2f""" % (storage.name,
                          capacity,
                          charge_capacity,
                          discharge_capacity,
                          storage_cost)
    system_cost += storage_cost
    system_co2 += co2

  print 'SYSTEM_COST: $%.2f' % system_cost
  print 'SYSTEM_CO2: %.2f Tonnes' % system_co2


def main():

  profiles_path = get_data_directory() + ['profiles', 'profiles_california.csv']
  profiles_file = osp.join(*profiles_path)

  profiles = pd.read_csv(profiles_file, index_col=0, parse_dates=True)
  lp = simple_lp(profiles)
  adjust_lp_policy(lp)

  if not lp.solve():
    raise ValueError("""LP did not converge.
Failure to solve is usually because of high RPS and no storage.""")

  display_lp_results(lp)


if __name__ == '__main__':
  main()


gridsim/grid_sim_website_example.py
# Copyright 2017 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


"""Partial Example of using grid_sim_linear_program to match website results.

First time users are encouraged to look at grid_sim_simple_example.py
first.

Analyzes energy similarly to how it was done for the backend to the
website http://energystrategies.org.  Note that this only calculates
for one out of the thirteen regions classified by the EIA.  To match
the results from the site, you must sum the results from all 13
regions as defined in REGION_HUBS.keys()

This backend uses a linear program to optimize the cheapest way to
generate electricity subject to certain constraints.  The constraints
are:
  1. Hourly generation must always exceed demand.  AKA.  Keep the lights on.
  2. Solution must have some percentage [0, 100] of power from a
     'Renewable Portfolio Standard' (RPS).  Different RPS have
     different standards so the code allows the user to specify which
     sources to put in the RPS.

Other factors which affect the linear program output include:
  1. Source fixed and variable costs. ($ / MW and $ / MWH respectively)
  2. Source CO2 per Mwh if there is a Carbon Tax (Tonnes / MWh)
  3. Carbon Tax Value ($ / Tonne)
"""

import math

import os.path as osp

import grid_sim_linear_program as gslp
import grid_sim_simple_example as simple

import pandas as pd


# Where to base a HVDC line.
# key:region, value: (latitude, longitude) in decimal degrees
REGION_HUBS = {'southwest': (33.4484, 112.0740),  # Phoenix
               'midatlantic': (40.4406, 79.9959),  # Pittsburgh
               'tva': (36.1627, 86.7816),  # Nashville
               'midwest': (39.0997, 94.5786),  # Kansas City
               'nyiso': (40.7128, 74.0059),  # New York City
               'ercot': (30.2672, 97.7431),  # Austin
               'northwest': (47.6062, 122.3321),  # Seattle
               'central': (41.2524, 95.9980),  # Omaha
               'carolinas': (35.2271, 80.8431),  # Charlotte
               'southeast': (33.5207, 86.8025),  # Birmingham
               'florida': (33.5207, 86.8025),  # Orlando
               'neiso': (33.5207, 86.8025),  # Boston
               'california': (34.0522, 118.2437)}  # Los Angeles


def circle_route_distance(coordinates1, coordinates2):
  """Calculate the spherical distance between two coordinates on earth.

  As described in https://en.wikipedia.org/wiki/Great-circle_distance

  Args:
    coordinates1: A tuple containint (latitude, longitude) in decimal
      degrees of point 1.
    coordinates2: A tuple containint (latitude, longitude) in decimal
      degrees of point 2.

  Returns:
    Distance on earth in kilometers between the two points.
  """

  lat1, lon1 = coordinates1
  lat2, lon2 = coordinates2

  # Convert to radians.
  lat1 *= math.pi / 180.
  lon1 *= math.pi / 180.
  lat2 *= math.pi / 180.
  lon2 *= math.pi / 180.

  delta_lat = abs(lat2 - lat1)
  delta_lon = abs(lon2 - lon1)

  # Result angle (asin) cannot be computed if point is exactly on
  # opposite side of earth, so special case out that point.  All other
  # values can be correctly calculated with the haversine, even values
  # close to 0, pi.
  if delta_lat == math.pi and delta_lon == math.pi:
    result_angle = math.pi
  else:
    haversine_lat = math.sin(delta_lat / 2) ** 2
    haversine_lon = math.sin(delta_lon / 2) ** 2
    result_angle = 2 * math.asin(math.sqrt(
        haversine_lat + math.cos(lat1) * math.cos(lat2) * haversine_lon))

  earth_radius_km = 6371
  return earth_radius_km * result_angle


def get_transmission_cost_efficiency(coordinates1, coordinates2):
  """Return transmission capital cost $/MW and transmission efficiency.

  Args:
    coordinates1: A tuple containint (latitude, longitude) in decimal
      degrees of point 1.
    coordinates2: A tuple containint (latitude, longitude) in decimal
      degrees of point 2.

  Returns:
    A tuple containing (cost $/MW, efficiency) elements.
  """

  # Assumptions: HVDC is 97% efficient / 1000 km with
  # Costs: $334 / km / MWh

  distance_km = circle_route_distance(coordinates1, coordinates2)
  capital_cost_per_mw = distance_km * 334

  efficiency = pow(0.97, distance_km / 1000)

  return (capital_cost_per_mw, efficiency)


def configure_sources_and_storage(profile_directory,
                                  region,
                                  source_dataframe,
                                  storage_dataframe,
                                  source_dict_index,
                                  storage_names,
                                  rps_names,
                                  hydrolimits=None):
  """Generates a LinearProgramContainer similar to the website.

  Args:
    profile_directory: String filepath where profile files of the form
      profile_<region>.csv exist.  Acceptable values of <region> are
      REGION_HUBS.keys().

    region: String name of the region to generate simulation for.
      Acceptable values are REGION_HUBS.keys().

    source_dataframe: A pandas dataframe with source names in indexed
      column-0.

      Index names match the source names in the website with a cost
      option index. (e.g. COAL_0, SOLAR_2, NUCLEAR_1).  Some sources
      only have one cost option associated with them.
      Source names are as follows.
        COAL, Coal fired plants.
        HYDROPOWER, Water turbines fed by a dammed reservoir.
        NGCC, Natural Gas Combined Cycle. A natural gas fired turbine
          whose exhaust heats a steam turbine for additional power.
        NGCT, Natural Gas Conventional Turbine. A natural gas turbine.
        NGCC_AMINE, NGCC + Carbon Capture and Sequestration via amine
          capture.
        NGCC_CRYO, NGCC + Carbon Capture and Sequestration from
          capturing the CO2 by freezing into dry-ice.
        NUCLEAR, Nuclear power.
        SOLAR, Solar power through utility scale photovoltaic panels.
        WIND, Wind power through modern conventional wind turbines.

      Column headers include
        CO2, Amount of CO2 emitted per generated energy. Units are
          Tonnes of CO2 Emitted per Megawatt-hour.
        fixed, Cost of building and maintaining a plant per
          nameplate-capacity of the plant.  Units are $ / Megawatt

        variable, Cost of running the plant based upon generation
          energy.  Units are $ / Megawatt-hour

    storage_dataframe: A pandas dataframe with storage names in indexed
      column-0.

      There are only two kinds of storage considered in this dataframe.
        HYDROGEN, Storage charges by generating and storing hydrogen.
          Storage discharges by buring hydrogen.
        ELECTROCHEMICAL, Storage charges and discharges through batteries.

      Index names match the storage names.

      Column headers include:
        fixed,
        charge_efficiency,
        discharge_efficiency,
        charge_capital,
        discharge_capital,

    source_dict_index: A dict keyed by source name with index values
      for cost.  Acceptable index values are either [0] for sources
      which only had one cost assumed in the webiste or [0,1,2] for
      sources which had 3 choices for cost assumption.

    storage_names: A list of storge names of storage type to add to
      the LP.  Must match names in the storage_dataframe index.

    rps_names: A list of source names which should be considered in
      the Renewable Portfolio Standard.

    hydrolimits: A dict with 'max_power' and 'max_energy' keys.  If
      specified, hydropower will be limited to this power and energy.

  Returns:
    A Configured LinearProgramContainer suitable for simulating.
  """
  profiles_file = osp.join(profile_directory, 'profiles_usa.csv')

  profiles_dataframe = pd.read_csv(profiles_file, index_col=0, parse_dates=True)

  lp = gslp.LinearProgramContainer(profiles_dataframe)

  # Specify grid load or demand which has a profile in
  # profile_dataframe.<region>_DEMAND
  lp.add_demands(gslp.GridDemand('%s_DEMAND' % region.upper()))

  # Configure dispatchable and non-dispatchable sources.
  for source_name, source_index in source_dict_index.iteritems():
    dataframe_row = source_dataframe.loc['%s_%d' % (source_name, source_index)]
    is_rps_source = source_name in rps_names

    # Adjust source_name by region to match profiles and to
    # differentiate it from sources from other regions.
    regional_source_name = '%s_%s' % (region.upper(), source_name)
    source = gslp.GridSource(name=regional_source_name,
                             nameplate_unit_cost=dataframe_row['fixed'],
                             variable_unit_cost=dataframe_row['variable'],
                             co2_per_electrical_energy=dataframe_row['CO2'],
                             is_rps_source=is_rps_source)

    # For energystrategies.org we assumed that the prime hydropower
    # sites have already been developed and built.  So for hydropower
    # sites, we make the capital cost 0.  Here we limit the LP to only
    # use as much power and energy as existing sites already provide.
    # Without this limitation and with capital cost of 0, the LP will
    # assume an infinite supply of cheap hydropower and fulfill demand
    # with 100% hydropower.

    if hydrolimits is not None:
      if source_name == 'HYDROPOWER':
        source.max_power = hydrolimits['max_power']
        source.max_energy = hydrolimits['max_energy']

    # Non-dispatchable sources have profiles associated with them.
    if regional_source_name in profiles_dataframe.columns:
      lp.add_nondispatchable_sources(source)
    else:
      lp.add_dispatchable_sources(source)

  # Add Solar and Wind from other regions.  Adjust by additional
  # transmission costs and efficiency losses for distance traveled if
  # Solar and Wind are in the simulation.  The website assumes Solar
  # and Wind are always present.

  for other_region in REGION_HUBS:
    if other_region != region:
      transmission_cost, efficiency = get_transmission_cost_efficiency(
          REGION_HUBS[region],
          REGION_HUBS[other_region])

      for source in ['SOLAR', 'WIND']:
        if source in source_dict_index:
          cost_row = source_dataframe.loc['%s_%d'
                                          % (source,
                                             source_dict_index[source])]
          lp.add_nondispatchable_sources(
              gslp.GridSource(
                  name='%s_%s' % (other_region.upper(), source),
                  nameplate_unit_cost=cost_row['fixed'] + transmission_cost,
                  variable_unit_cost=cost_row['variable'],
                  co2_per_electrical_energy=cost_row['CO2'],
                  is_rps_source=True,
                  power_coefficient=efficiency)
          )

  for storage_name in storage_names:
    dataframe_row = storage_dataframe.loc[storage_name]
    storage = gslp.GridRecStorage(
        name=storage_name,
        storage_nameplate_cost=dataframe_row['fixed'],
        charge_nameplate_cost=dataframe_row['charge_capital'],
        discharge_nameplate_cost=dataframe_row['discharge_capital'],
        charge_efficiency=dataframe_row['charge_efficiency'],
        discharge_efficiency=dataframe_row['discharge_efficiency'])

    lp.add_storage(storage)

  return lp


def main():

  region = 'california'

  data_dir = simple.get_data_directory()
  source_cost_path = data_dir + ['costs', 'source_costs.csv']
  storage_cost_path = data_dir + ['costs', 'storage_costs.csv']
  hydrolimits_path = data_dir + ['costs', 'regional_hydro_limits.csv']
  profile_path = data_dir + ['profiles']

  source_costs_file = osp.join(*source_cost_path)
  storage_costs_file = osp.join(*storage_cost_path)
  hydro_limits_file = osp.join(*hydrolimits_path)

  source_costs_dataframe = pd.read_csv(source_costs_file, index_col=0)
  storage_costs_dataframe = pd.read_csv(storage_costs_file, index_col=0)
  hydrolimits_dataframe = pd.read_csv(hydro_limits_file, index_col=0)

  ng_cost_index = 0
  cost_settings = {
      'COAL': 0,
      'HYDROPOWER': 0,
      'NGCC': ng_cost_index,
      'NGCT': ng_cost_index,
      'NGCC_CRYO': ng_cost_index,
      'WIND': 2,
      'SOLAR': 2,
      'NUCLEAR': 2
  }

  storage_names = ['ELECTROCHEMICAL']
  rps_names = ['SOLAR', 'WIND']

  profile_directory = osp.join(*profile_path)
  lp = configure_sources_and_storage(
      region=region,
      profile_directory=profile_directory,
      source_dataframe=source_costs_dataframe,
      storage_dataframe=storage_costs_dataframe,
      source_dict_index=cost_settings,
      storage_names=storage_names,
      rps_names=rps_names,
      hydrolimits=hydrolimits_dataframe.loc[region]
  )

  simple.adjust_lp_policy(
      lp,
      carbon_tax=50,  # $50 per tonne
      renewable_portfolio_percentage=20,  # 20% generated from rps_names
      annual_discount_rate=0.06,  # 6% annual discount rate.
      lifetime_in_years=30)  # 30 year lifetime

  print 'Solving may take a few minutes...'
  if not lp.solve():
    raise ValueError("""LP did not converge.
Failure to solve is usually because of high RPS and no storage.""")

  simple.display_lp_results(lp)


if __name__ == '__main__':
  main()
